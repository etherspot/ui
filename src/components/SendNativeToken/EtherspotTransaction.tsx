/* MIT License
 *
 * Copyright (c) 2024 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotTransaction } from '@etherspot/transaction-kit';
import { isEmpty } from 'lodash';

// Utils
import { nativeTokenValidationMessage } from '../../utils/validation';
import { parseBigNumber } from '../../utils/bigNumber';

// Types
import type { SendNativeTokenTransactionProps } from '../../models/Transactions';

/**
 * @name EtherspotTokenTransaction
 * @description: The following <EtherspotTokenTransaction />
    component will transfer [value] [token] from the built-in
    Etherspot Wallet account to the receiverAddress.
 * @param {SendNativeTokenTransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function EtherspotTokenTransaction(props: SendNativeTokenTransactionProps) {
  if (!isEmpty(nativeTokenValidationMessage(props)) || !parseBigNumber(props.value)?.gt('0.0')) return null;

  return (
    <EtherspotTransaction
      to={props.receiverAddress}
      value={props.value}
    />
  );
}

export default EtherspotTokenTransaction;
