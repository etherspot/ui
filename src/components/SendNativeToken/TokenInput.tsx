/* MIT License
 *
 * Copyright (c) 2024 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React, { useState } from 'react';

// Utils
import { normalizeRawValue, parseBigNumber } from '../../utils/bigNumber';
import { numberInProgressRegex } from '../../utils/common';
import { nativeTokenValidationMessage } from '../../utils/validation';

// Types
import type { SendNativeTokenInputProps, SendNativeTokenTransactionProps } from '../../models/Transactions';

/**
 * @name TokenInput
 * @description: Use of NativeTokenInput as token value input.
 * @param {SendNativeTokenInputProps & SendNativeTokenTransactionProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
function TokenInput(props: SendNativeTokenInputProps & SendNativeTokenTransactionProps) {
  const { value, errorMessageClassName, onChangeValue, decimals, className, disableSendOnEnter, handleEnterPress } =
    props;
  const [errorMessage, setErrorMessage] = useState('');

  // Function is called when the input value changes 
  const onChange = (event) => {
    const { value: newValue } = event.target;

    // Normalize the value and check if it is a valid number
    const normalizedValue = normalizeRawValue(newValue, decimals);
    if (!numberInProgressRegex.test(normalizedValue)) return;

    // Parse the normalized value into a BigNumber
    const parseValue = parseBigNumber(normalizedValue);
    if (parseValue?.isNaN()) return;

    // Check receiverAddress and value is valid or not.
    setErrorMessage(nativeTokenValidationMessage({ ...props, value: normalizedValue }));

    // Call the onChangeValue prop with the normalized value
    onChangeValue && onChangeValue(normalizedValue);
    
  };

  // Function is called when a key is pressed in the input field
  const onKeyDown = () => {
    // If disableSendOnEnter is false, call the handleEnterPress prop
    if (!disableSendOnEnter) handleEnterPress && handleEnterPress();
  };

  return (
    <>
      <input
        type="text"
        name="value"
        id={`etherspot-ui-token-${Math.random().toString()}`}
        value={value}
        onChange={onChange}
        className={className}
        onKeyDown={onKeyDown}
        placeholder="0"
      />
      <p className={errorMessageClassName}>{errorMessage}</p>
    </>
  );
}

export default TokenInput;
