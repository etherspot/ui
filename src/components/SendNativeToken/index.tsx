/* MIT License
 * 
 * Copyright (c) 2023 Etherspot
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
import React, {useRef} from 'react';

// Components
import SendNativeTokenUI from './SendNativeTokenUI';

// Services
import {
  EtherspotTransactionKit
} from "@etherspot/transaction-kit";

// Props interface for the SendNativeToken component
/**
 * @typedef {Object} SendNativeTokenProps - Props for the SendNativeToken component.
 * @property {string} receiverAddress - The Ethereum address of the receiver.
 * @property {number} chain - The Ethereum chain ID.
 * @property {string} [className] - Additional tailwind CSS class for styling.
 * @property {React.CSSProperties} [style] - Inline styles for the component overwrite tailwind CSS.
 * @property {boolean} [unstyled=false] - Flag indicating whether to apply default styles.
 * @property {boolean} [debug=false] - Flag indicating whether to enable debugging.
 * @property {React.ReactNode | React.RefObject<HTMLElement>} [triggerElement] - Trigger element for UI interactions.
 * @property {boolean} [disableSendOnEnter=false] - Flag indicating whether to disable sending on Enter key press.
 * @property {boolean} [onlyEstimate=false] - Flag indicating whether to only estimate the transaction.
 * @property {Object | undefined} [provider] - Ethereum provider (random wallet provider).
 */

// Props interface for the SendNativeToken component
interface SendNativeTokenProps {
  receiverAddress: string;
  chain: number;
  className?: string;
  style?: React.CSSProperties;
  unstyled?: boolean;
  debug?: boolean;
  triggerElement?: React.ReactNode | React.RefObject<HTMLElement>;
  disableSendOnEnter?: boolean;
  onlyEstimate?: boolean;
  provider?: Object | undefined;
}

/**
 * @name SendNativeToken
 * @description SendNativeTokenUI component provides a UI for sending native crypto tokens
 * to another Ethereum address using the EtherspotTransactionKit.This user-friendly interface 
 * is designed to simplify the complex underlying transaction mechanics, making it accessible 
 * to users who may not be familiar with the intricacies of blockchain transactions.
 * @param {SendNativeTokenProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */

// SendNativeToken component
const SendNativeToken = ({
  receiverAddress,
  chain,
  className,
  style,
  unstyled = false,
  debug = false,
  triggerElement,
  disableSendOnEnter = false,
  onlyEstimate = false,
  provider
}: SendNativeTokenProps) => {
  
  return (
    // EtherspotTransactionKit component, wrapping the SendNativeTokenUI, 
    // Use EtherspotTransactionKit as wrapper from this file and import SendNativeToken from different file, otherwise transaction kit will not find parent provider.
    <EtherspotTransactionKit
      provider={provider}
      chainId={chain}
    >
      {/* Rendering the SendNativeTokenUI component */}
     <SendNativeTokenUI
        receiverAddress={receiverAddress}
        chain={chain} 
        className= {className}
        unstyled={unstyled} 
        debug={debug}
        disableSendOnEnter={disableSendOnEnter}
        onlyEstimate={onlyEstimate}
        style={style}
        triggerElement={triggerElement}
      />
    </EtherspotTransactionKit>
  );
}

export default SendNativeToken;



