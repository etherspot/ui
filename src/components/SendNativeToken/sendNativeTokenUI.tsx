import React, { useState, KeyboardEvent, useEffect } from 'react';
import {
  EtherspotBatches,
  EtherspotBatch,
  EtherspotTransaction,
  useEtherspotTransactions,
} from "@etherspot/transaction-kit";
import { ethers } from "ethers";
import { isValidEthereumAddress } from '../../utils/validation';


interface SendNativeTokenProps {
  receiverAddress: string;
  chain: number;
  className?: string;
  style?: React.CSSProperties;
  unstyled?: boolean;
  debug?: boolean;
  triggerElement?: React.ReactNode | React.RefObject<HTMLElement>;
  disableSendOnEnter?: boolean;
  onlyEstimate?: boolean;
}

const SendNativeTokenUI = ({
  receiverAddress,
  chain,
  className,
  style,
  unstyled = false,
  debug = false,
  triggerElement,
  disableSendOnEnter = false,
  onlyEstimate = false,
}: SendNativeTokenProps) => {

  const [amount, setAmount] = useState('0.001');
  const [isValidAddress, setIsValidAddress] = useState(true);
  const [error, setError] = useState('');
  const { estimate, send } = useEtherspotTransactions();
  const randomWallet = ethers.Wallet.createRandom();
  const providerWallet = new ethers.Wallet(randomWallet.privateKey);
  console.log(providerWallet);

  useEffect(() => {
    // Add event listener for triggerElement click
    const triggerElementRef = triggerElement as React.MutableRefObject<HTMLElement> | undefined;
    if (triggerElementRef && triggerElementRef.current) {
      triggerElementRef.current.addEventListener('click', estimateAndSend);
    }

    // Cleanup
    return () => {
      if (triggerElementRef && triggerElementRef.current) {
        triggerElementRef.current.removeEventListener('click', estimateAndSend);
      }
    };
  }, [triggerElement]);

  useEffect(() => {
    // validate receiver address
    if(isValidEthereumAddress(receiverAddress)){  
      setIsValidAddress(true);
    }else{
      setIsValidAddress(false);
    }
  }, [receiverAddress]);

  // handle on key press logic of sending 
  const handleEnterPress = (event: KeyboardEvent<HTMLInputElement>) => {
    if (!disableSendOnEnter && event.key === 'Enter') {
      // Trigger sending logic here
      estimateAndSend()
    }
  };

  // on change logic
  const handleAmountChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setAmount(event.target.value);
  };

  // logic of estimate and send based on props
  /*const estimateAndSend = async () => {
    if (isValidAddress) {
      // Estimate logic
      const gasEstimation = await transactionKitProps.provider.estimateGas({
        to: transactionKitProps.to,
        value: transactionKitProps.amount,
      });
  
      if (!onlyEstimate) {
        // Send logic
        try {
          const transactionHash = await transactionKitProps.provider.sendTransaction({
            to: transactionKitProps.to,
            value: transactionKitProps.amount,
            gasLimit: gasEstimation.toNumber() + 10000, // Adding some buffer to the estimated gas limit
          });
  
          console.log('Transaction sent! Transaction Hash:', transactionHash);
          // Optionally, you can handle the success of the transaction here
        } catch (error) {
          console.error('Error sending transaction:', error);
          // Optionally, you can handle the failure of the transaction here
        }
      }
  
      if (debug) {
        console.log('send');
      }
    } else {
      setError('Receiver address is not a valid blockchain address');
    }
  }; */

  const estimateAndSend = async () => {
    if (isValidAddress) {
      // Estimate logic
      const gasEstimation = await estimate();
  
      if (!onlyEstimate) {
       send();
      }
  
      if (debug) {
        console.log('send');
      }
    } else {
      setError('Receiver address is not a valid blockchain address');
    }
  };
  
  const transactionKitProps = {
    provider: providerWallet,
    to: receiverAddress,
    chainId: chain,
    amount: onlyEstimate ? 0 : parseFloat(amount),
    debug,
  };

  return (
      <EtherspotBatches>
        <EtherspotBatch>
          <EtherspotTransaction to={transactionKitProps.to} value={transactionKitProps.amount}>
            <div className="mt-2.5">
              <input
                type="text"
                value={amount}
                onChange={handleAmountChange}
                onKeyDown={handleEnterPress}
                placeholder="Enter Amount"
                className= {unstyled ? '' : className }
                style={style}
              />
              {error ? <p className='text-sm'>{error}</p> : ''}
            </div>
          </EtherspotTransaction>
        </EtherspotBatch>
      </EtherspotBatches>
  );
  
};

export default SendNativeTokenUI;