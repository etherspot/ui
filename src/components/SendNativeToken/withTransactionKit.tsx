/* MIT License
 *
 * Copyright (c) 2024 Etherspot
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from 'react';
import { EtherspotBatch, EtherspotBatches, EtherspotTransactionKit } from '@etherspot/transaction-kit';

// Types
import { SendNativeTokenProps } from '../../models/Transactions';

// Required Props for the SendNativeToken component
/**
 * @typedef {Object} SendNativeTokenProps - Props for the SendNativeTokenProps component.
 * @property {string} receiverAddress - The Ethereum address of the receiver.
 * @property {string} value - value of Native token.
 * @property {string} onChangeValue - gets updated value from onChangeValue.
 * @property {number} chainId - The Ethereum chain ID.
 * @property {ethers.Wallet} provider - The Ethereum provider.
 * @property {number} decimals - decimals for Native token(using for convert number to bignumber).
 */

/**
 * @name SendNativeToken
 * @description Using SendNativeToken component to we can send token. 
    For sending token need tokenAddress, receiverAddress and value.
 * @param {SendNativeTokenProps} props - The props for the component
 * @returns {React.ReactElement} The rendered component.
 */
const withTransactionKit = (WrappedComponent) => {
  // This is a higher-order component that wraps the provided component with the EtherspotTransactionKit and related components
  const TransactionKit = (props: SendNativeTokenProps) => {
    const { chainId, provider } = props;

    return (
      <EtherspotTransactionKit provider={provider} >
        <EtherspotBatches>
          <EtherspotBatch chainId={chainId}>
            <WrappedComponent {...props} />
          </EtherspotBatch>
        </EtherspotBatches>
      </EtherspotTransactionKit>
    );
  };
  return TransactionKit;
};

export default withTransactionKit;
