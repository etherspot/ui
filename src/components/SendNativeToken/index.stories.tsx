import React from 'react';
import { Meta, Story } from '@storybook/react';
import SendNativeToken  from './index';

export default {
  title: 'Components/SendNativeToken',
  component: SendNativeToken,
} as Meta;

const Template: Story<SendNativeTokenProps> = (args) => <SendNativeToken {...args} />;

export const NativeToken = Template.bind({});
NativeToken.args = {
  receiverAddress: '0x725404c8Eead111d9E6DFE118c535F43402a9511',
  chain: 1,
  className: 'block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6',
  unstyled: false,
  debug: true,
  disableSendOnEnter: false,
  onlyEstimate: false,
};


