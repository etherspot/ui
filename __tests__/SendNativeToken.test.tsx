import {render, fireEvent, cleanup } from '@testing-library/react';
import SendNativeToken from '../src/components/SendNativeToken/index';

afterEach(cleanup);
test('Send native token renders correctly', () => {
  const tree = render(<SendNativeToken
    receiverAddress="0x725404c8Eead111d9E6DFE118c535F43402a9511"
    chain={1} 
    className= 'block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6'
    unstyled={false} 
    debug={true}
    disableSendOnEnter={false}
    onlyEstimate={false}
  />);
  expect(tree).toMatchSnapshot();
});

test('input value updates correctly', () => {
  // Arrange
  const { getByPlaceholderText } = render(<SendNativeToken
    receiverAddress="0x725404c8Eead111d9E6DFE118c535F43402a9511"
    chain={1} 
    className= 'block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6'
    unstyled={false} 
    debug={true}
    disableSendOnEnter={false}
    onlyEstimate={false}
  />
  );

  // Act: Change the input value
  const inputElement = getByPlaceholderText('Enter Amount');
  fireEvent.change(inputElement, { target: { value: '0.001' } });

  // Assert: Check if the input value has been updated
  expect(inputElement.value).toBe('0.001');
});

test('input value updates correctly', () => {
  // Arrange
  const { getByPlaceholderText } = render(<SendNativeToken
    receiverAddress="0x725404c8Eead111d9E6DFE118c535F43402a9511"
    chain={1} 
    className= 'block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6'
    unstyled={false} 
    debug={true}
    disableSendOnEnter={false}
    onlyEstimate={false}
  />
  );

  // Act: Change the input value
  const inputElement = getByPlaceholderText('Enter Amount');
  fireEvent.keyDown(inputElement, { target: { value: '0.001' } });

  // Assert: call function of send
  expect(inputElement.value).toBe('0.001');
});