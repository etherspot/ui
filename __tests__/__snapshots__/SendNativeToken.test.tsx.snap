// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send native token renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mt-2.5"
      >
        <input
          class="block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
          placeholder="Enter Amount"
          type="text"
          value="0.001"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mt-2.5"
    >
      <input
        class="block w-1/2 rounded-md border-1 border-gray-300 px-3.5 py-2 text-gray-900 shadow-sm ring-1 ring-inset ring-gray-600 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-indigo-600 sm:text-sm sm:leading-6"
        placeholder="Enter Amount"
        type="text"
        value="0.001"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
